HDDS-3208. Implement Ratis snapshot on SCM

amaliujia opened the pull request:
## What changes were proposed in this pull request?

Design doc: https://docs.google.com/document/d/1uy4_ER2V6nNQJ7_5455Wz8NmI142JHPnif6Y1OdPi8E/edit?usp=sharing

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3208

## How was this patch tested?

UT

amaliujia commented:
I found one of the failed test is `TestMiniOzoneCluster.testDNstartAfterSCM`, which failed on loading RocksDB when restarting SCM. Will try investigate the root cause.

amaliujia commented:
cc @nandakumar131 @ChenSammi @GlenGeng

GlenGeng commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java:
```diff
@@ -101,6 +102,16 @@ public ContainerManagerImpl(
       final PipelineManager pipelineManager,
       final Table<ContainerID, ContainerInfo> containerStore)
       throws IOException {
+    this(conf, scmHaManager, pipelineManager, containerStore, null);
```
Could we avoid using this Ctor ? We may need to change the related test cases.

amaliujia replied:
Same https://issues.apache.org/jira/browse/HDDS-4634.

This is because MockHAManager does not have a mock buffer implementation.

GlenGeng replied:
Could we implement the mock buffer in this PR ? Since missing this change will pollute the production code, and will also give a burden to the on-going PR of deleted block log.

GlenGeng commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/metadata/SCMTransactionInfoCodec.java:
```diff
@@ -0,0 +1,48 @@
+  @Override
+  public byte[] toPersistedFormat(SCMTransactionInfo object)
+      throws IOException {
+    checkNotNull(object, "Null object can't be converted to byte array.");
```
Ditto,  could we reuse `OMTransactionInfoCodec` ?

GlenGeng commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineManagerV2Impl.java:
```diff
@@ -114,11 +115,21 @@ public static PipelineManagerV2Impl newPipelineManager(
       ConfigurationSource conf, SCMHAManager scmhaManager,
       NodeManager nodeManager, Table<PipelineID, Pipeline> pipelineStore,
       EventPublisher eventPublisher) throws IOException {
+    return newPipelineManager(conf, scmhaManager, nodeManager, pipelineStore,
```
Ditto, may we remove this Ctor ?

GlenGeng commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMDBTransactionBuffer.java:
```diff
@@ -0,0 +1,70 @@
+        SCMTransactionInfo
+            .builder()
+            .setTransactionIndex(-1)
+            .setCurrentTerm(-1)
```
`.setCurrentTerm(0)`. The first leader will be elected from term 1.

GlenGeng reviewed (commented):
LGTM. only some inline comments.

Could we rename the title to `implement Ratis takeSnapshot on SCM`?

amaliujia commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMTransactionInfo.java:
```diff
@@ -0,0 +1,111 @@
+
+import static org.apache.hadoop.ozone.OzoneConsts.TRANSACTION_INFO_SPLIT_KEY;
+
+final public class SCMTransactionInfo {
```
To reuse `OMTransactionInfo`, my current answer is probably no.

Though both `OMTransactionInfo` and `SCMTransactionInfo` seems are doing the same thing for now, decoupling is good when there is a need that either `OMTransactionInfo` and `SCMTransactionInfo` need to add new/drop existing fields while the other does not need. Especially in this time that SCM snapshot is still under development. 

I can create a JIRA though to track the merge of two transaction info.

runzhiwang replied:
Sorry, I'm not clear why we can not reuse OMTransactionInfo. They has the same fields.

amaliujia replied:
I think we are better to reuse `OMTransactionInfo` after merge HDDS-2823 back to master.

Not reusing `OMTransactionInfo` is to try to reduce the risk that have conflicts after syncing master int HDDS-2823. Also, to reuse `OMTransactionInfo`, I think we'd better to update the name to `TransactionInfo`, which will be a source of complicates, unless we can change it in the master branch first.

To conclude a bit, we can reuse it after merging SCM HA back to master, the goal is try to reduce potential conflicts if there could be any.

SCM HA is in developing branch anyway and I am sure we will need to clean up lots of stuff eventually. As long as we have JIRAs to not lose track of such work.

GlenGeng replied:
`OMTransactionInfo` is written by Bharat, since SCM and OM has a quite similar requirement, we may have a talk with him, to determine whether OM and SCM can share the same `TransactionInfo` and the `Codec`. @amaliujia  What do you think ?

amaliujia replied:
Sure we can do that. I will need add some changes in main branch, and then sync that to HDDS-2823 and then reuse in SCM.

I think we should do that separately as that will take time. I have created https://issues.apache.org/jira/browse/HDDS-4660 to track this effort.

amaliujia commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerStateManagerImpl.java:
```diff
@@ -275,12 +278,24 @@ public void addContainer(final ContainerInfoProto containerInfo)
+              transactionBuffer.getCurrentBatchOperation(),
+              containerID, container);
+        } else {
+          containerStore.put(containerID, container);
```
This is leave for MockHAManager code path in which the buffer is NULL :) 

If we choose to update MockHAManager to have a sort of MockBuffer implementation, then we can move this `else branch` because at that time I think all existing code will access this buffer.

I can create a JIRA to track that add buffer to MockHAManager and remove this else :-)

amaliujia replied:
https://issues.apache.org/jira/browse/HDDS-4634

Create JIRA to add buffer in MockHAManager thus we can remove this `null` check.

amaliujia commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:
```diff
@@ -89,4 +98,12 @@ private Message process(final SCMRatisRequest request)
 
+  @Override
+  public long takeSnapshot() throws IOException {
+    LOG.info("Current Snapshot Index {}", getLastAppliedTermIndex());
```
makes sense.

amaliujia replied:
Done

amaliujia commented:
With https://github.com/apache/ozone/pull/1733 we will be able to write tests for snapshot based on on single Ratis server SCM setup in MiniOzoneCluster.

ChenSammi commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMDBTransactionBuffer.java:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  public void updateLatestTrxInfo(SCMTransactionInfo info) {
+    this.latestTrxInfo = info;
```
Shall we add some checks here to guarantee new info is greater than current lastTrxInfo?

amaliujia replied:
Good point. I will add a check or a LOG.error for that case.

Theoretically, the Ratis log will be applied with index monotonically increasing thus the new info is supposed to greater than current  lastTrxInfo. But have a check or a LOG will better handle rare cases or bugs.

amaliujia replied:
Done

bshashikant replied:
In my opinion, we should stop the ratis server here of such a case arise. It is fatal case where we cannot move ahead from here.

amaliujia replied:
Such case should be considered as a bug and be fixed. 

I created a JIRA to further discuss how to deal with such case: https://issues.apache.org/jira/browse/HDDS-4723

ChenSammi commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/metadata/SCMDBDefinition.java:
```diff
@@ -93,6 +104,6 @@ public String getLocationConfigKey() {
   @Override
   public DBColumnFamilyDefinition[] getColumnFamilies() {
     return new DBColumnFamilyDefinition[] {DELETED_BLOCKS, VALID_CERTS,
-        REVOKED_CERTS, PIPELINES, CONTAINERS};
```
Since we add a new table in SCM DB,  we need a JIRA to track how to handle back compatibility during upgrade non HA SCM to HA SCM.

amaliujia replied:
https://issues.apache.org/jira/browse/HDDS-4635

ChenSammi commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/metadata/SCMMetadataStore.java:
```diff
@@ -110,6 +111,11 @@
+  /**
+   *
+   */
+  Table <String, SCMTransactionInfo> getTransactionInfoTable();
```
Are we missing some comments here?

amaliujia commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:
```diff
@@ -89,4 +98,12 @@ private Message process(final SCMRatisRequest request)
+  @Override
+  public long takeSnapshot() throws IOException {
+    LOG.info("Current Snapshot Index {}", getLastAppliedTermIndex());
+    TermIndex lastTermIndex = getLastAppliedTermIndex();
```
I believe we don't need to. 

```
    public TermIndex getLastAppliedTermIndex() {
        return (TermIndex)this.lastAppliedTermIndex.get();
    }
```

```
    protected boolean updateLastAppliedTermIndex(long term, long index) {
        TermIndex newTI = TermIndex.newTermIndex(term, index);
        TermIndex oldTI = (TermIndex)this.lastAppliedTermIndex.getAndSet(newTI);
         ....
```

Both function access `this.lastAppliedTermIndex`.

ChenSammi replied:
Since all rocksdb writes are batched to write to disk in takesnapshot,  we need to change takesnapshot default frequency of ratis during scm ratis server creation.

amaliujia replied:
> Since all rocksdb writes are batched to write to disk in takesnapshot, we need to change takesnapshot default frequency of ratis during scm ratis server creation.

https://issues.apache.org/jira/browse/HDDS-4636 I will prefer to handle separately.

ChenSammi commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/metadata/SCMDBDefinition.java:
```diff
@@ -80,6 +82,15 @@
+  public static final DBColumnFamilyDefinition<String, SCMTransactionInfo>
+      TRANSACTIONINFO =
+      new DBColumnFamilyDefinition<>(
+          "scmTransactionInfoTable",
```
scmTransactionInfoTable -> scmTransactionInfos

ChenSammi commented:
@amaliujia, thanks for working on this,  I left a few comments.

amaliujia commented:
@ChenSammi  thank you! I will start to address comments since there are a bunch of actionable ones existing.

linyiqun commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMDBTransactionBuffer.java:
```diff
@@ -0,0 +1,75 @@
+
+import static org.apache.hadoop.ozone.OzoneConsts.TRANSACTION_INFO_KEY;
+
+public class SCMDBTransactionBuffer {
```
Could we document javadoc for this? That will be easier understood by others.

linyiqun commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMTransactionInfo.java:
```diff
@@ -0,0 +1,119 @@
+
+import static org.apache.hadoop.ozone.OzoneConsts.TRANSACTION_INFO_SPLIT_KEY;
+
+final public class SCMTransactionInfo {
```
Also would be great to add javadoc for this class.

runzhiwang commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:
```diff
@@ -89,4 +99,15 @@ private Message process(final SCMRatisRequest request)
+    transactionBuffer.flush();
+    LOG.debug("SCM takeSnapshot: {} took {} ms",
+        Time.monotonicNow() - startTime);
+    return lastAppliedIndex;
```
Maybe flush should return SCMTransactionInfo, and this return SCMTransactionInfo#transactionIndex ?

amaliujia replied:
Code changes a bit. Maybe take another look?

runzhiwang commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMDBTransactionBuffer.java:
```diff
@@ -0,0 +1,75 @@
+            .build();
+  }
+
+  public BatchOperation getCurrentBatchOperation() {
```
Expose currentBatchOperation has a risk, if someone commit this batch operation not by calling flush(), then applyIndex did not wrote into RocksDB.

amaliujia replied:
hmm I think the goal here is commiting batch operations won't need to flush DB. The flushing is controlled by takeSnapshot().

Is there a case that committing will have a following flush?

runzhiwang replied:
I'm afraid if someone do not know must call flush() to sync, but sync `currentBatchOperation `directly, then inconsistent will happen.

amaliujia replied:
Do you know is there a better mechanism to achieve the buffering? 

I studied what OM does and that is different from SCM. OM just buffer entry and then apply entries in a batch, but SCM need to router such entry to handlers and then different handler will apply the entry.

So if we change to the way that OM is doing, there will be a good amount of refactoring needs happen, which might be not appropriate in this PR. E.g. We will need to move handlers to the buffer class and insert entries into buffer. Buffer will be the place to trouter entries to right state managers.

What do you think? I can create a PR to track such refactoring thus we won't need to expose  `currentBatchOperation`

GlenGeng replied:
How about return an anonymous subclass of `RDBBatchOperation` whose 'commit()' will throw `RuntimeException`, which means you can only write to the batch, but there is no way to commit this batch ?

`RDBBatchOperation#commit()` is called by `RDBStore#commitBatchOperation()`

amaliujia replied:
Glen's idea might be much easier. That's a good option.

I think I will separate it to:
1. add javadoc to `getCurrentBatchOperation()` to remind that do not use this batch to commit.
2. Apply a solution to fix this issue. Created https://issues.apache.org/jira/browse/HDDS-4661 to track it.

amaliujia commented:
@ChenSammi @runzhiwang @GlenGeng @linyiqun 

I have rebased and updated this PR. Now this PR includes the takeSnapshot and loadSnapshot with a test.

amaliujia commented:
Will fix failing tests.

linyiqun commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMRatisSnapshotInfo.java:
```diff
@@ -0,0 +1,76 @@
+  public void updateTermIndex(long newTerm, long newIndex) {
+    this.term = newTerm;
+    this.snapshotIndex = newIndex;
+  }
```
I don't see this method is used in this PR. Will this be used in follow-up task? If not we could remove this and remove volatile keyword for term/snapshotIndex as well. Because I didn't find the concurrent update for this two variables.

amaliujia replied:
got it. Will remove this method.

The `volatile` is a good point. I am also looking for comments to point me out whether there are potential concurrent operations.  I will remove `volatile` if there is no comment about this situation.

amaliujia replied:
That makes sense. Done

GlenGeng reviewed (commented):
Thanks for the job!

Better solve following task in this PR, it should be part of the job of load snapshot.
https://issues.apache.org/jira/browse/HDDS-4533

amaliujia commented:
> Thanks for the job!
> 
> Better solve following task in this PR, it should be part of the job of load snapshot.
> https://issues.apache.org/jira/browse/HDDS-4533

@GlenGeng  I prefer to fix such issue separately so create a PR for it: https://github.com/apache/ozone/pull/1796

amaliujia commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:
```diff
@@ -89,4 +128,33 @@ private Message process(final SCMRatisRequest request)
+    transactionBuffer.flush();
+    transactionBuffer.setLatestSnapshot(
+        transactionBuffer.getLatestTrxInfo().toSnapshotInfo());
+    LOG.debug("SCM takeSnapshot: {} took {} ms",
```
Done

amaliujia commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:
```diff
@@ -55,15 +84,25 @@ public void registerHandler(RequestType type, Object handler) {
           Message.valueOf(trx.getStateMachineLogEntry().getLogData()));
-      applyTransactionFuture.complete(process(request));
+      applyTransactionFuture.complete(
+          process(
```
+1

amaliujia commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMDBTransactionBuffer.java:
```diff
@@ -0,0 +1,101 @@
+
+    metadataStore.getStore().commitBatchOperation(currentBatchOperation);
+    currentBatchOperation.close();
+
```
The snapshot info is set also during the initialization. So not only `flush()` will need to update that information.

amaliujia commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:
```diff
@@ -55,15 +84,25 @@ public void registerHandler(RequestType type, Object handler) {
-  private Message process(final SCMRatisRequest request)
-      throws Exception {
+  private boolean shouldUpdate(SCMTransactionInfo info) {
+    return !(info.getTransactionIndex() == -1 && info.getTerm() == 0);
```
Makes sense

GlenGeng commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineManagerV2Impl.java:
```diff
@@ -111,15 +113,25 @@ private PipelineManagerV2Impl(ConfigurationSource conf,
   }
 
-  public static PipelineManagerV2Impl newPipelineManager(
+  public static PipelineManagerV2Impl newPipelineManagerWithMockBuffer(
```
`DBTransactionBuffer` can be fetched by calling `SCMHAManager#getDBTransactionBuffer`, so this function `newPipelineManagerWithMockBuffer` is not needed.
```
// Create PipelineStateManager
StateManager stateManager = PipelineStateManagerV2Impl
    .newBuilder().setPipelineStore(pipelineStore)
    .setRatisServer(scmhaManager.getRatisServer())
    .setNodeManager(nodeManager)
    .setSCMDBTransactionBuffer(scmhaManager.getDBTransactionBuffer())
    .build()
```

amaliujia replied:
This is a very good point.

amaliujia commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMTransactionInfo.java:
```diff
@@ -0,0 +1,145 @@
+    this.transactionIndex = transactionIndex;
+  }
+
+  public boolean shouldUpdate() {
```
`isInitialized` is a good naming.

amaliujia commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:
```diff
@@ -54,7 +54,7 @@ public SCMStateMachine(SCMDBTransactionBuffer buffer) throws SCMException {
     this.transactionBuffer = buffer;
     SCMTransactionInfo latestTrxInfo = buffer.getLatestTrxInfo();
-    if (latestTrxInfo.shouldUpdate()) {
+    if (!latestTrxInfo.isInitialized()) {
```
Actually it is `!latestTrxInfo.isInitialized()`.

`SCMTransactionInfo latestTrxInfo = buffer.getLatestTrxInfo();` comes from the buffer. The buffer will load the SCMTransactionInfo from the DB. If there was no snapshot taken before (e.g. brand new SCM), there isn't a SCMTransactionInfo from DB thus buffer will create a default one. 

So we only need to updateLastAppliedTermIndex when latestTrxInfo is not initialized (thus it is loaded from DB and it's meaningful).

bshashikant commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:
```diff
@@ -89,4 +119,30 @@ private Message process(final SCMRatisRequest request)
+    transactionBuffer.flush();
+    LOG.info("Current Snapshot Index {}, takeSnapshot took {} ms",
+        getLastAppliedTermIndex(), Time.monotonicNow() - startTime);
+    return lastAppliedIndex;
```
why the lastApplied index returned here is the one derived from transaction buffer? Once the snapshot is taken, the last applied index should be given by getLastAppliedTermIndex() not transaction buffer which may be behind actual last applied index in ratis?

amaliujia replied:
O the code here is keeping the lastAppliedIndex in buffer to equal to the getLastAppliedTermIndex() actually. 

```
    LOG.info("Current Snapshot Index {}, takeSnapshot took {} ms",
        getLastAppliedTermIndex(), Time.monotonicNow() - startTime);
```
is wrong.

I should replace `getLastAppliedTermIndex()` with `lastAppliedIndex` to make it clear.


We need the `lastAppliedIndex` in buffer keep updated so that will be written into DB.

bshashikant replied:
I see the buffer getting updated with applied index only after applyTransaction is called. ApplyTransaction is not invoked for every log transaction in Ratis. So, its better to rely on getLastAppliedTermIndex() rather than the buffer for the latest applied index.

bshashikant commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineManagerV2Impl.java:
```diff
@@ -114,12 +114,14 @@ private PipelineManagerV2Impl(ConfigurationSource conf,
       ConfigurationSource conf, SCMHAManager scmhaManager,
       NodeManager nodeManager, Table<PipelineID, Pipeline> pipelineStore,
-      EventPublisher eventPublisher) throws IOException {
+      EventPublisher eventPublisher)
```
unintended change??

amaliujia replied:
Yes :) will undo it.

amaliujia commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMDBTransactionBuffer.java:
```diff
@@ -0,0 +1,112 @@
+
+    metadataStore.getStore().commitBatchOperation(currentBatchOperation);
+    currentBatchOperation.close();
+    this.latestSnapshot = latestTrxInfo.toSnapshotInfo();
```
I agree with it :-)

In fact we have discussed not to expose batch operation above and I created  https://issues.apache.org/jira/browse/HDDS-4661.

As this PR becomes larger and larger so I am planning to address this batch operation related comments in HDDS-4661, including when to init and close it.

How do you feel? Do you agree with my idea?

bshashikant replied:
I am ok with addressing in a different PR.

amaliujia commented:
Failed UT is the decommission one, which is known to be flaky.

bshashikant commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:
```diff
@@ -26,27 +26,55 @@
+  }
+
   @Override
   public CompletableFuture<Message> applyTransaction(
```
Can we rename applyTransaction to applyTansactionSerial as this is a serialized operation anyways?

amaliujia replied:
I created https://issues.apache.org/jira/browse/HDDS-4684.

After check the Ratis state machine interface, I can see two functions:
```
  /**
   * Called for transactions that have been committed to the RAFT log. This step is called
   * sequentially in strict serial order that the transactions have been committed in the log.
   * The SM is expected to do only necessary work, and leave the actual apply operation to the
   * applyTransaction calls that can happen concurrently.
   * @param trx the transaction state including the log entry that has been committed to a quorum
   *            of the raft peers
   * @return The Transaction context.
   */
  TransactionContext applyTransactionSerial(TransactionContext trx);

  /**
   * Apply a committed log entry to the state machine. This method can be called concurrently with
   * the other calls, and there is no guarantee that the calls will be ordered according to the
   * log commit order.
   * @param trx the transaction state including the log entry that has been committed to a quorum
   *            of the raft peers
   */
  CompletableFuture<Message> applyTransaction(TransactionContext trx);
```

So there is no API as the following 
```
  Message applyTransactionSerial(TransactionContext trx);
```

Basically a function to apply transaction in strict serial order where it returns a Message.

I am planning to send an email to Ratis community to discuss the intention of only having one applyTransaction that returns Message (CompletableFuture<Message>) but can be called concurrently.

We can address this API change in HDDS-4684.

amaliujia commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java:
```diff
@@ -397,4 +396,8 @@ public void close() throws IOException {
     return containerStateManager;
   }
+
+  public SCMHAManager getSCMHAManager() {
```
Done

bshashikant reviewed (approved):
The patch looks ok. The changes discussed will be addressed in subsequent jira.

The transaction buffer flush now can only happen via a ratis snapshot , but in case if ratis is not enabled, there needs to be rocks db sync on every update to db or we need a way to periodically flush the buffer changes to db. This can be done as a part of separate jira.

bshashikant commented:
Thanks @amaliujia for the contribution.
